core/domain/exp_services_test.py

```diff
@@ -3057,3 +3069,131 @@ def test_reject_actioned_suggestion(self):
             with self.assertRaisesRegexp(Exception, exception_message):
                 exp_services.reject_suggestion(
                     self.editor_id, self.THREAD_ID1, self.EXP_ID2)
+
+
+class EditorAutoSavingUnitTests(test_utils.GenericTestBase):
+    """Test editor auto saving functions in exp_services."""
+
+    THREAD_ID1 = '1111'
+    EXP_ID1 = 'exp_id1'
+    EXP_ID2 = 'exp_id2'
+    EXP_ID3 = 'exp_id3'
+    USERNAME = 'user123'
+    USER_ID = 'user_id'
+    COMMIT_MESSAGE = 'commit message'
+    DATETIME = datetime.datetime.strptime('2016-02-16', '%Y-%m-%d')
+    OLDER_DATETIME = datetime.datetime.strptime('2016-01-16', '%Y-%m-%d')
+    NEWER_DATETIME = datetime.datetime.strptime('2016-03-16', '%Y-%m-%d')
+    DRAFT_CHANGELIST = {'new_content': {}}
+    NEW_CHANGELIST = [{
+        'cmd': 'edit_exploration_property',
+        'property_name': 'title',
+        'new_value': 'New title'}]
+
+    def setUp(self):
+        super(EditorAutoSavingUnitTests, self).setUp()
+        self.current_exp_id = None
+        # Explorations with draft set.
+        user_models.ExplorationUserDataModel(
+            id='%s.%s' % (self.USER_ID, self.EXP_ID1), user_id=self.USER_ID,
+            exploration_id=self.EXP_ID1,
+            draft_change_list=self.DRAFT_CHANGELIST,
+            draft_change_list_last_updated=self.DATETIME,
+            draft_change_list_exp_version=3).put()
+        user_models.ExplorationUserDataModel(
+            id='%s.%s' % (self.USER_ID, self.EXP_ID2), user_id=self.USER_ID,
+            exploration_id=self.EXP_ID2,
+            draft_change_list=self.DRAFT_CHANGELIST,
+            draft_change_list_last_updated=self.DATETIME,
+            draft_change_list_exp_version=4).put()
+        # Exploration with no draft.
+        user_models.ExplorationUserDataModel(
+            id='%s.%s' % (self.USER_ID, self.EXP_ID3), user_id=self.USER_ID,
+            exploration_id=self.EXP_ID3).put()
+
+    def _get_exploration_summary(self, unused_exp_id):
+        return exp_models.ExpSummaryModel(version=3)
+
+    def _empty_method(self, strict):
+        pass
+
+    def _get_exp_domain_object(self, exp_id, unused_change_list):
+        self.current_exp_id = exp_id
+        return exp_domain.Exploration.create_default_exploration(
+            self.EXP_ID1, 'A title', 'A category')
+
+    def test_draft_version_valid_true(self):
+        with self.swap(exp_services, 'get_exploration_summary_by_id',
```

Comment:
I think it's better if get_exploration_summary_by_id and get_exploration_by_id are the real thing. Either that or the mock functions should return real things (but I think the former is easier). At the moment it all feels a little fragile, since the summaries and change lists don't seem to be valid.

It might be a good idea to populate the datastore with some explorations in setUp(), if needed.